File: src/Bisync.Api/Services/ProductionCostAttribution.cs
namespace Bisync.Api.Services;

/// <summary>
/// Splits production batch cost across primary + bi-product outputs.
/// Locked rule: bi unit = base × (pct/100); primary residual share / primary qty.
/// Example: base 10, total 20, primary 10, bi 10 @ 50% → primary 15, bi 5.
/// When every bi line is 100%, split batch total equally across all outputs.
/// </summary>
public static class ProductionCostAttribution
{
    public sealed record BiLine(string Key, decimal Quantity, decimal AttributionPct);

    public sealed record BiLineResult(string Key, decimal Quantity, decimal AttributionPct, decimal UnitCost, decimal Share);

    public sealed record Result(
        decimal BatchTotalCost,
        decimal PrimaryQty,
        decimal PrimaryUnitCost,
        decimal PrimaryShare,
        IReadOnlyList<BiLineResult> BiLines);

    public static Result Allocate(
        decimal baseUnitCost,
        decimal totalQty,
        decimal primaryQty,
        IReadOnlyList<BiLine> biLines)
    {
        baseUnitCost = baseUnitCost < 0 ? 0 : baseUnitCost;
        totalQty = totalQty < 0 ? 0 : totalQty;
        primaryQty = primaryQty < 0 ? 0 : primaryQty;
        var lines = (biLines ?? Array.Empty<BiLine>())
            .Where(l => l.Quantity > 0)
            .ToList();
        var batchTotal = baseUnitCost * totalQty;

        if (totalQty <= 0)
        {
            return new Result(
                0,
                primaryQty,
                0,
                0,
                lines.Select(l => new BiLineResult(l.Key, l.Quantity, l.AttributionPct, 0, 0)).ToList());
        }

        var allHundred = lines.Count > 0
            && lines.All(l => Math.Abs(l.AttributionPct - 100m) < 0.0001m);

        if (allHundred)
        {
            var outputs = new List<(string Key, decimal Qty)>();
            if (primaryQty > 0) outputs.Add(("primary", primaryQty));
            outputs.AddRange(lines.Select(l => (l.Key, l.Quantity)));
            var n = Math.Max(1, outputs.Count);
            var shareEach = batchTotal / n;
            var primaryShare = primaryQty > 0 ? shareEach : 0;
            return new Result(
                batchTotal,
                primaryQty,
                primaryQty > 0 ? primaryShare / primaryQty : 0,
                primaryShare,
                lines.Select(l =>
                {
                    var share = l.Quantity > 0 ? shareEach : 0;
                    return new BiLineResult(
                        l.Key,
                        l.Quantity,
                        l.AttributionPct,
                        l.Quantity > 0 ? share / l.Quantity : 0,
                        share);
                }).ToList());
        }

        var attributed = lines.Select(l =>
        {
            var pct = l.AttributionPct < 0 ? 0 : l.AttributionPct;
            var unit = baseUnitCost * (pct / 100m);
            var share = unit * l.Quantity;
            return new BiLineResult(l.Key, l.Quantity, pct, unit, share);
        }).ToList();
        var biShare = attributed.Sum(l => l.Share);
        var primaryShareResidual = Math.Max(0, batchTotal - biShare);
        var primaryUnit = primaryQty > 0 ? primaryShareResidual / primaryQty : 0;
        return new Result(batchTotal, primaryQty, primaryUnit, primaryShareResidual, attributed);
    }
}

File: client/src/data/productionCostAttribution.ts
/**
 * Production cost attribution for B2B Principal / Sub-Product + Bi-Product outputs.
 *
 * Batch total B = baseUnitCost × totalQty.
 *
 * Equal split (every attributionPct === 100):
 *   share_j = B / nOutputs, unit_j = share_j / qty_j
 *
 * Weighted bi attribution (normal):
 *   unit_bi_i = baseUnitCost × (attributionPct_i / 100)
 *   share_bi_i = unit_bi_i × qty_i
 *   share_primary = B − Σ share_bi  (≥ 0)
 *   unit_primary = share_primary / primaryQty
 *
 * Locked example: base 10, total 20, primary 10, bi 10 @ 50%
 *   → bi unit 5, primary unit 15.
 */

export type AttributionOutputLine = {
  key: string;
  quantity: number;
  attributionPct: number;
};

export type AttributionResult = {
  batchTotalCost: number;
  primaryQty: number;
  primaryUnitCost: number;
  primaryShare: number;
  biLines: {
    key: string;
    quantity: number;
    attributionPct: number;
    unitCost: number;
    share: number;
  }[];
};

export function allocateProductionCost(input: {
  baseUnitCost: number;
  totalQty: number;
  primaryQty: number;
  biLines: AttributionOutputLine[];
}): AttributionResult {
  const baseUnitCost = Number.isFinite(input.baseUnitCost) ? Math.max(0, input.baseUnitCost) : 0;
  const totalQty = Number.isFinite(input.totalQty) ? Math.max(0, input.totalQty) : 0;
  const primaryQty = Number.isFinite(input.primaryQty) ? Math.max(0, input.primaryQty) : 0;
  const biLines = (input.biLines ?? []).filter(line => line.quantity > 0);
  const batchTotalCost = baseUnitCost * totalQty;

  if (totalQty <= 0 || batchTotalCost < 0) {
    return {
      batchTotalCost: 0,
      primaryQty,
      primaryUnitCost: 0,
      primaryShare: 0,
      biLines: biLines.map(line => ({
        key: line.key,
        quantity: line.quantity,
        attributionPct: line.attributionPct,
        unitCost: 0,
        share: 0,
      })),
    };
  }

  const allHundred = biLines.length > 0
    && biLines.every(line => Math.abs(line.attributionPct - 100) < 0.0001);

  if (allHundred) {
    const outputs = [
      { key: 'primary', quantity: primaryQty },
      ...biLines.map(line => ({ key: line.key, quantity: line.quantity })),
    ].filter(line => line.quantity > 0);
    const n = outputs.length || 1;
    const shareEach = batchTotalCost / n;
    const primaryShare = primaryQty > 0 ? shareEach : 0;
    return {
      batchTotalCost,
      primaryQty,
      primaryUnitCost: primaryQty > 0 ? primaryShare / primaryQty : 0,
      primaryShare,
      biLines: biLines.map(line => {
        const share = line.quantity > 0 ? shareEach : 0;
        return {
          key: line.key,
          quantity: line.quantity,
          attributionPct: line.attributionPct,
          unitCost: line.quantity > 0 ? share / line.quantity : 0,
          share,
        };
      }),
    };
  }

  const attributedBi = biLines.map(line => {
    const pct = Number.isFinite(line.attributionPct) ? Math.max(0, line.attributionPct) : 0;
    const unitCost = baseUnitCost * (pct / 100);
    const share = unitCost * line.quantity;
    return {
      key: line.key,
      quantity: line.quantity,
      attributionPct: pct,
      unitCost,
      share,
    };
  });
  const biShareTotal = attributedBi.reduce((sum, line) => sum + line.share, 0);
  const primaryShare = Math.max(0, batchTotalCost - biShareTotal);
  const primaryUnitCost = primaryQty > 0 ? primaryShare / primaryQty : 0;

  return {
    batchTotalCost,
    primaryQty,
    primaryUnitCost,
    primaryShare,
    biLines: attributedBi,
  };
}
